soroban-env-common/env.json

```diff
@@ -1560,6 +1560,61 @@
+                },
+                {
+                    "export": "1",
+                    "name": "call_reentrant",
+                    "args": [
+                        {
+                            "name": "contract",
+                            "type": "AddressObject"
+                        },
+                        {
+                            "name": "func",
+                            "type": "Symbol"
+                        },
+                        {
+                            "name": "args",
+                            "type": "VecObject"
+                        }
+                    ],
+                    "return": "Val",
+                    "docs": "Calls a function in another contract with arguments contained in vector `args`. If the call is successful, returns the result of the called function. Traps otherwise. This functions enables re-entrancy in the immediate cross-contract call.",
+                    "min_supported_protocol": 21
+                },
+                {
+                    "export": "2",
+                    "name": "try_call_reentrant",
+                    "args": [
+                        {
+                            "name": "contract",
+                            "type": "AddressObject"
+                        },
+                        {
+                            "name": "func",
+                            "type": "Symbol"
+                        },
+                        {
+                            "name": "args",
+                            "type": "VecObject"
+                        }
+                    ],
+                    "return": "Val",
+                    "docs": "Calls a function in another contract with arguments contained in vector `args`, returning either the result of the called function or an `Error` if the called function failed. The returned error is either a custom `ContractError` that the called contract returns explicitly, or an error with type `Context` and code `InvalidAction` in case of any other error in the called contract (such as a host function failure that caused a trap). `try_call` might trap in a few scenarios where the error can't be meaningfully recovered from, such as running out of budget. This functions enables re-entrancy in the immediate cross-contract call.",
+                    "min_supported_protocol": 21
+                },
+                {
+                    "export": "3",
+                    "name": "set_reentrant",
+                    "args": [
+                        {
+                            "name": "enabled",
+                            "type": "Bool"
+                        }
+                    ],
+                    "return": "Void",
+                    "docs": "Enables the current contract to specify the reentrancy rules.",
+                    "min_supported_protocol": 21
```

Comment:
Is the idea with `set_reentrant` that it provides an alternative way to enable reentry on all the following regular `call` / `try_call` host fns?

Reply:
Set reentrant just instructs the host that the contract in this context (or better, currently it sets the flag on the whole host). The current implementation requires two specialized host functions to have the callparam option with the reentry enabled (while call and try_call always have it disabled), and `set_reentrant` is an additional guard to make sure that if it's set to false `call_reentrant` will fail. So `set_reentrant(true)` and `call` in the current implementation would result in a trap anyways. This would change if we inherit the reentry guard from the context i.e what dmytro brought up in the discord.